hoot-core/src/main/cpp/hoot/core/visitors/ImplicitTaggerBase.cpp

```diff
@@ -0,0 +1,616 @@
+          matchingWords);
+        tagsToAdd = implicitlyDerivedTags;
+      }
+      else
+      {
+        //we didn't find any tags for the whole names, so let's look for them with the tokenized
+        //name parts
+        QStringList nameTokensList = _getNameTokens(filteredNames);
+
+        LOG_VARD(nameTokensList);
+        LOG_VARD(implicitlyDerivedTags.size());
+        LOG_VARD(nameTokensList.size());
+
+        //only going up to token group size = 2, as larger sizes weren't found experimentally to
+        //yield any better results
+        if (implicitlyDerivedTags.size() == 0 && nameTokensList.size() > 2)
+        {
+          _getImplicitlyDerivedTagsFromMultipleNameTokens(
+            filteredNames, nameTokensList, e->getTags(), implicitlyDerivedTags, matchingWords,
+            wordsInvolvedInMultipleRules);
+        }
+
+        if (implicitlyDerivedTags.size() == 0)
+        {
+          //didn't find any matches with the token groups, so let's try with single tokens
+          _getImplicitlyDerivedTagsFromSingleNameTokens(
+            filteredNames, nameTokensList, e->getTags(), implicitlyDerivedTags, matchingWords,
+            wordsInvolvedInMultipleRules, namesContainBuilding, namesContainOffice);
+        }
+        LOG_VARD(implicitlyDerivedTags);
+        LOG_VARD(matchingWords);
+        LOG_VARD(wordsInvolvedInMultipleRules);
+
+        if (wordsInvolvedInMultipleRules)
+        {
+          LOG_DEBUG(
+            "Found duplicate match for name tokens: " << nameTokensList << " with matching words: " <<
+            matchingWords);
+          foundDuplicateMatch = true;
+        }
+        else if (implicitlyDerivedTags.size() > 0)
+        {
+          LOG_DEBUG(
+            "Derived implicit tags for name tokens: " << nameTokensList << " with matching words: " <<
+            matchingWords);
+          tagsToAdd = implicitlyDerivedTags;
+        }
+      }
+      LOG_VARD(tagsToAdd);
+
+      LOG_VARD(foundDuplicateMatch);
+      if (foundDuplicateMatch)
+      {
+        _updateElementForDuplicateMatch(e, matchingWords);
+      }
+      else if (!tagsToAdd.isEmpty())
+      {
+        assert(!matchingWords.isEmpty());
+        LOG_VART(matchingWords);
+
+        _ensureCorrectTagSpecificity(e, tagsToAdd);
+
+        //This is a little kludgy, but we'll leave it for now since it helps.
+        if (tagsToAdd.isEmpty() && !_elementIsASpecificPoi && namesContainOffice)
+        {
+          tagsToAdd.appendValue("building", "office");
+          matchingWords.insert("office");
+        }
+        else if (tagsToAdd.isEmpty() && !_elementIsASpecificPoi && namesContainBuilding)
+        {
+          tagsToAdd.appendValue("building", "yes");
+          matchingWords.insert("building");
+        }
+
+        if (!tagsToAdd.isEmpty())
+        {
+          _addImplicitTags(e, tagsToAdd, matchingWords);
+        }
+      }
+    }
+
+    _numNodesParsed++;
+    if (_numNodesParsed % 1000 == 0)
```

Comment:
I think this should be
`_numNodesParsed % _statusUpdateInterval == 0`

Reply:
I think that I had wanted something less than _statusUpdateInterval but shouldn't have hardcoded it.  _statusUpdateInterval / 10 should work.